File: DesktopCore/src/screen/DebugConsole.java
package screen;

import java.awt.BorderLayout;
import java.awt.Color;

import javax.swing.JFrame;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.text.DefaultCaret;

import util.DebugOutput;

public class DebugConsole extends JFrame
						  implements DebugOutput
{
	private boolean scrollLock = false;
	
	public DebugConsole()
	{
		getContentPane().setLayout(new BorderLayout(0, 0));
		getContentPane().add(scrollPane);
		
		textArea.setForeground(Color.GREEN);
		textArea.setBackground(Color.BLACK);
		textArea.setEditable(false);
		textArea.setLineWrap(true);
		scrollPane.setViewportView(textArea);
		
		DefaultCaret caret = (DefaultCaret)textArea.getCaret();
		caret.setUpdatePolicy(DefaultCaret.NEVER_UPDATE);
	}
	
	private final JScrollPane scrollPane = new JScrollPane();
	private final JTextArea textArea = new JTextArea();
	
	public void scrollToBottom()
	{
		if (!scrollLock)
		{
			String allText = textArea.getText();
			int length = allText.length();
		
			textArea.setCaretPosition(length);
		}
	}
	
	public void setScrollLock(boolean scrollLock)
	{
		this.scrollLock = scrollLock;
	}

	@Override
	public void append(String text)
	{
		textArea.append(text);
		scrollToBottom();
	}

	@Override
	public String getLogs()
	{
		return textArea.getText();
	}
	
	@Override
	public void clear()
	{
		textArea.setText("");
	}
}

File: Darts/src/main/kotlin/dartzee/core/screen/DebugConsole.kt
package dartzee.core.screen

import dartzee.core.util.Debug
import dartzee.core.util.DebugOutput
import java.awt.BorderLayout
import java.awt.Color
import javax.swing.JFrame
import javax.swing.JScrollPane
import javax.swing.JTextPane
import javax.swing.text.BadLocationException
import javax.swing.text.DefaultStyledDocument
import javax.swing.text.StyleConstants
import javax.swing.text.StyleContext

class DebugConsole : JFrame(), DebugOutput
{
    private val doc = DefaultStyledDocument()
    private val scrollPane = JScrollPane()
    private val textArea = JTextPane(doc)

    init
    {
        title = "Console"
        setSize(1000, 600)
        setLocationRelativeTo(null)
        contentPane.layout = BorderLayout(0, 0)
        contentPane.add(scrollPane)
        textArea.foreground = Color.GREEN
        textArea.background = Color.BLACK
        textArea.isEditable = false
        scrollPane.setViewportView(textArea)
    }

    override fun append(text: String)
    {
        val cx = StyleContext()
        val style = cx.addStyle(text, null)
        if (text.contains(Debug.SQL_PREFIX))
        {
            when
            {
                text.contains("INSERT") -> StyleConstants.setForeground(style, Color.ORANGE)
                text.contains("UPDATE") -> StyleConstants.setForeground(style, Color.ORANGE)
                text.contains("DELETE") -> StyleConstants.setForeground(style, Color.RED)
                else -> StyleConstants.setForeground(style, Color.CYAN)
            }
        }

        try
        {
            doc.insertString(doc.length, text, style)
            textArea.select(doc.length, doc.length)
        }
        catch (ble: BadLocationException)
        {
            Debug.stackTrace(ble, "BLE trying to append: $text")
        }
    }

    override fun getLogs(): String
    {
        return try
        {
            doc.getText(0, doc.length)
        }
        catch (t: Throwable)
        {
            Debug.stackTrace(t)
            ""
        }
    }

    override fun clear()
    {
        textArea.text = ""
    }
}